### PR TITLE
Artifacts

### PR DESCRIPTION
**Pull request by adrianliechti**

### CODE REVIEW

File: artifacts/pkg/server/server_artifacts.go [hunk 1]
```diff
@@ -1,9 +1,28 @@
...
 
 func (s *Server) Artifacts(req *proto.ArtifactsRequest, server proto.Artifacts_ArtifactsServer) error {
+	artifacts, err := s.artifacts.Artifacts(server.Context(), req.Namespace, &artifacts.ArtifactListOptions{})
+
+	if err != nil {
+		return err
+	}
+
+	for _, a := range artifacts {
+		response := proto.ArtifactResponse{
```
Comment (adrianliechti) [inline artifacts/pkg/server/server_artifacts.go:16]: you might want to extract this to a "convert.go" file:
ArtifactResponse -> Artifact
Artifact -> ArtifactResponse

also helps others to understand where to centrally add fields

File: artifacts/artifacts.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+import "context"
+
+type Provider interface {
+	Artifacts(ctx context.Context, namespace string, options *ArtifactListOptions) ([]Artifact, error)
+}
+
+type ArtifactKind string
+
+type Artifact struct {
```
Comment (adrianliechti) [inline artifacts/artifacts.go:11]: imagine you build hundreds of artifacts...
over time, disk storage grows and grows and the api get slower and slower

let's help the customer to cleanup. can you explore, if it's possible to get data about:
- createdAt
- updatedAt
- downloadCount

File: artifacts/provider/gitlab/gitlab.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import (
+	"context"
+	"errors"
+
+	"github.com/platformplane/platform/artifacts"
+	"github.com/xanzy/go-gitlab"
+)
+
+const (
+	stringEmpty string = ""
```
Comment (adrianliechti) [inline artifacts/provider/gitlab/gitlab.go:12]: this is more on the untypical side in go: most packages compare "" directly for better readability and no unexpected side effects (alternative len(...) == 0)

if you need pointer magic, consider a ToStringPtr(string) / toString(*string) method (the gitlab client provides one). also this one is a good one as inpsiration: https://github.com/Azure/go-autorest/tree/main/autorest/to

File: artifacts/provider/gitlab/gitlab.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	client *gitlab.Client
+}
+
+type Config struct {
+	BaseUrl string
+	Token   string
+}
+
+func NewGitLabClient(config Config) (*GitLabClient, error) {
+	if config.BaseUrl == stringEmpty {
```
Comment (adrianliechti) [inline artifacts/provider/gitlab/gitlab.go:25]: you might like this package here to validate structs
https://github.com/go-playground/validator

also his package for environment variables can verify needed env variables:
https://github.com/kelseyhightower/envconfig

struct Config {
}

func (c *Config) Validate() error {
}

func (c *Config) MustValidate() {

log.Fatal(...)
}

File: artifacts/provider/gitlab/packages.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+		return nil, err
+	}
+
+	artifactsArray := make([]artifacts.Artifact, 0, len(packages))
+
+	for _, p := range packages {
+
+		name := p.Name + "@" + p.Version
+
+		x := artifacts.Artifact{
```
Comment (adrianliechti) [inline artifacts/provider/gitlab/packages.go:24]: use a convert helper function here

File: artifacts/provider/gitlab/containers.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	artifactsArray := make([]artifacts.Artifact, 0, len(repositories))
+
+	for _, rr := range repositories {
+
+		name := rr.Location
+		if rr.TagsCount > 0 {
```
Comment (adrianliechti) [inline artifacts/provider/gitlab/containers.go:23]: you might want to directly check the "Tags" slice and do not trust this field (could be misleading / end in panic if the api allows paging etc. and the field is bigger than the array)

File: artifacts/pkg/server/server_artifacts.go (hunk 1 above)
Comment (linkdotnet) [reply artifacts/pkg/server/server_artifacts.go]: Added a mapper

File: artifacts/provider/gitlab/gitlab.go (hunk 3 above)
Comment (linkdotnet) [reply artifacts/provider/gitlab/gitlab.go]: Removed this in favor of a validator.

File: artifacts/provider/gitlab/gitlab.go (hunk 4 above)
Comment (linkdotnet) [reply artifacts/provider/gitlab/gitlab.go]: Used validator

File: artifacts/provider/gitlab/packages.go (hunk 5 above)
Comment (linkdotnet) [reply artifacts/provider/gitlab/packages.go]: done

File: artifacts/provider/gitlab/containers.go (hunk 6 above)
Comment (linkdotnet) [reply artifacts/provider/gitlab/containers.go]: done

File: artifacts/artifacts.go (hunk 2 above)
Comment (linkdotnet) [reply artifacts/artifacts.go:14]: `createdAt` added

Comment (linkdotnet) [reply artifacts/artifacts.go:14]: I don't think "updatedAt" would make sense. Artifacts are somewhat immutable - so `createdAt == updatedAt` in virtually all the cases.

Comment (adrianliechti) [reply artifacts/artifacts.go:14]: in docker it is pretty common:

examples
- latest
- stable
- golang:1 (latest 1.xx release)
- debian:buster (that release with all the latest patches)

or images following semver...

Comment (linkdotnet) [reply artifacts/artifacts.go:14]: Sure - but every of those new images is a new unique artifact with its own id and whatnot.

An already created image does not get updated - so once we have an image with an Id "1" this "1" will stay the same over its lifetime

Comment (linkdotnet) [reply artifacts/artifacts.go:14]: That keeps me thinking. For "v2" of the API, where we want to handle docker/nuget special, we want to group all docker images with the same name.

**Pull request merged**